Lib/asyncio/taskgroups.py

```diff
@@ -54,21 +54,22 @@ async def __aenter__(self):
 
     async def __aexit__(self, et, exc, tb):
         self._exiting = True
-        propagate_cancellation_error = None
 
         if (exc is not None and
                 self._is_base_error(exc) and
                 self._base_error is None):
             self._base_error = exc
 
-        if et is not None:
-            if et is exceptions.CancelledError:
-                if self._parent_cancel_requested and not self._parent_task.uncancel():
-                    # Do nothing, i.e. swallow the error.
-                    pass
-                else:
-                    propagate_cancellation_error = exc
+        propagate_cancellation_error = \
+            exc if et is exceptions.CancelledError else None
+        if self._parent_cancel_requested:
+            # If this flag is set we *must* call uncancel().
+            if self._parent_task.uncancel() == 0:
+                # If there are no pending cancellations left,
+                # don't propagate CancelledError.
+                propagate_cancellation_error = None
```

Comment:
If we want to fix Thomas's example, we can probably fix it like this:
```suggestion
                propagate_cancellation_error = None
            elif et is not exceptions.CancelledError:
                # If the exception is not CancelledError,
                # replace it with CancelledError.
                propagate_cancellation_error = exceptions.CancelledError()
```

Reply:
Alas, not. It seems Thomas's example is not related to `uncancel()` -- that example doesn't even use `await` in the body of the `async with TaskGroup()` block, so `_parent_cancel_requested` is never set, and `uncancel()` *should* not be called.

That example needs to check whether one of the subtasks task was cancelled but not by us. There's no way we can fix that before RC1, but I also don't see that it is a fatal flaw -- it's a bug we can fix in 3.11.1 by adding more bookkeeping to `TaskGroup`.

Reply:
That can be dealt in another PR as "usual bug", it needs not to block this one.

Reply:
I've made a less dramatic version of what I think is this issue: 
https://gist.github.com/graingert/e8ddd68a0513bd7c02531a77a79cdacd